apps/service_providers/llm_service/runnables.py

```diff
@@ -227,11 +280,51 @@ def invoke(self, input: str, config: RunnableConfig | None = None) -> ChainOutpu
         if thread_id:
             input_dict["thread_id"] = thread_id
 
-        response: OpenAIAssistantFinish = assistant_runnable.invoke(input_dict, config)
+        response = self._get_response_with_retries(config, input_dict, thread_id)
         if not thread_id:
             self.chat.set_metadata(self.chat.MetadataKeys.OPENAI_THREAD_ID, response.thread_id)
 
         output = response.return_values["output"]
         self._save_message_to_history(output, ChatMessageType.AI)
 
         return ChainOutput(output=response.return_values["output"], prompt_tokens=0, completion_tokens=0)
+
+    def _get_response_with_retries(self, config, input_dict, thread_id):
+        assistant = self.experiment.assistant.get_assistant()
+        assistant_runnable = RunnableLambda(self.format_input) | assistant
+        for i in range(3):
+            try:
+                response: OpenAIAssistantFinish = assistant_runnable.invoke(input_dict, config)
+            except openai.BadRequestError as e:
+                self._handle_api_error(thread_id, assistant, e)
+            except ValueError as e:
+                if re.search(r"cancelling|cancelled", str(e)):
+                    raise GenerationCancelled(ChainOutput(output="", prompt_tokens=0, completion_tokens=0))
+            else:
+                return response
+        raise GenerationError("Failed to get response after 3 retries")
+
+    def _handle_api_error(self, thread_id, assistant, exc):
+        """Handle OpenAI API errors.
+        This should either raise an exception or return if the error was handled and the run should be retried.
+        """
+        message = exc.body.get("message") or ""
+        match = re.match(r".*(thread_[\w]+) while a run (run_[\w]+) is active.*", message)
+        if not match:
+            raise exc
+
+        error_thread_id, run_id = match.groups()
+        if error_thread_id != thread_id:
+            raise GenerationError(f"Thread ID mismatch: {error_thread_id} != {thread_id}", exc)
+
+        self._cancel_run(assistant, thread_id, run_id)
+
+    def _cancel_run(self, assistant, thread_id, run_id):
+        logger.info("Cancelling run %s in thread %s", run_id, thread_id)
+        run = assistant.client.beta.threads.runs.cancel(thread_id=thread_id, run_id=run_id)
+        cancelling = run.status == "cancelling"
+        while cancelling:
```

Comment:
Wondering if we should not include a timeout here, so incase something happens at openai and the status doesn't get updated, we don't go into an infinte loop. Thoughts?